torch/csrc/cuda/nccl.cpp

```diff
@@ -0,0 +1,175 @@
+#include "nccl.h"
+#include "torch/csrc/THP.h"
+#include "torch/csrc/Types.h"
+#include "torch/csrc/DynamicTypes.h"
+#include "torch/csrc/cuda/THCP.h"
+
+#include <nccl.h>
+#include <sstream>
+#include <unordered_map>
+
+static inline void CHECK(ncclResult_t status)
+{
+  if (status != ncclSuccess) {
+    std::stringstream err;
+    err << "NCCL Error " << status << ": " << ncclGetErrorString(status);
+    throw std::runtime_error(err.str());
+  }
+}
+
+// TODO: make this thread_local + add mutexes
+std::unordered_map<std::string, ncclComm_t*> _communicators;
+
+PyObject * THCPModule_nccl_destroy(PyObject *args) {
+  for(auto it: _communicators) {
+    ncclCommDestroy(*(it.second));
+  }
+  Py_RETURN_NONE;
+}
+
+static ncclComm_t* _get_communicator(std::vector<at::Tensor>& inputs, int ndevices) {
+  std::stringstream hash;
+  std::vector<int> devs;
+  for (int i = 0; i < ndevices; i++) {
+    int dev = inputs[i].get_device();
+    hash <<  dev << ",";
+    devs.push_back(dev);
+  }
+  if (_communicators.find(hash.str()) == _communicators.end()) {
+    ncclComm_t *comms = (ncclComm_t*) malloc(sizeof(ncclComm_t) * ndevices);
```

Comment:
just use `new`. it will throw upon failure. Unless nccl expects to steal your memory allocated by `malloc` and free it for you in `ncclCommDestroy`. Otherwise you have a memory leak

Reply:
it eventually calls `free` on comm, so i cant use `new`: https://github.com/NVIDIA/nccl/blob/7fef264bfa3fce60907b1cd6808257c64e222604/src/core.cu#L581

Reply:
That free is on an individual `ncclComm_t`, not the array of communicators. As paszke wrote above, you should use C++ for the memory management. Something like:

```c++
struct NcclCommList {
  std::unique_ptr<ncclComm_t[]> comms;
  NcclCommList(IntList devices) 
     : comms(new ncclComm_t[devices.size()]) {
        // call ncclCommInitAll
   }
  ~NcclCommList() {
       // ncclCommDestroy each ncclComm_t
  }
};
```

The value type in the map would then be `NcclCommList`.

Reply:
I think we should try to avoid variable length arrays (present in `new ncclComm_t[devices.size()]`), as [it might not be supported on MSVC](https://stackoverflow.com/questions/5246900/enabling-vlasvariable-length-arrays-in-ms-visual-c).

Reply:
It's not a variable length array (VLA), at least not as the term is used to refer to the GCC C extension. The `new[]` operator is part of standard C++.